qutip/solver/nonmarkov/transfertensor.py

```diff
@@ -0,0 +1,171 @@
+
+[1] Javier Cerrillo and Jianshu Cao, Phys. Rev. Lett 112, 110401 (2014)
+"""
+
+import numpy as np
+import time
+from qutip import spre, vector_to_operator, operator_to_vector, Result
+
+
+def ttmsolve(dynmaps, state0, times, e_ops=[], options=None):
+    """
+    Expand time-evolution using the Transfer Tensor Method [1]_, based on a set
+    of precomputed dynamical maps.
+
+    Parameters
+    ----------
+    dynmaps : list of :class:`qutip.Qobj`, callable
+        List of precomputed dynamical maps (superoperators) for the first times
+        of ``times`` or a callback function that returns the superoperator at a
+        given time.
+
+    state0 : :class:`qutip.Qobj`
+        Initial density matrix or state vector (ket).
+
+    times : array_like
+        List of times :math:`t_n` at which to compute results.
+        Must be uniformily spaced.
+
+    e_ops : :class:`qutip.qobj`, callable, or list.
+        Single operator or list of operators for which to evaluate
+        expectation values or callable or list of callable.
+        Callable signature must be, `f(t: float, state: Qobj)`.
+        See :func:`expect` for more detail of operator expectation.
+
+    options : dictionary
+        Dictionary of options for the solver.
+
+        - store_final_state : bool
+          Whether or not to store the final state of the evolution in the
+          result class.
+        - store_states : bool, None
+          Whether or not to store the state vectors or density matrices.
+          On `None` the states will be saved if no expectation operators are
+          given.
+        - normalize_output : bool
+          Normalize output state to hide ODE numerical errors.
+        - threshold : float
+          Threshold for halting. Halts if  :math:`||T_{n}-T_{n-1}||` is below
+          treshold.
+        - num_learning : int
+          Number of times used to construct the dynmaps operators when
+          ``dynmaps`` is a callable.
+
+    Returns
+    -------
+    output: :class:`qutip.solver.Result`
+        An instance of the class :class:`qutip.solver.Result`.
+
+    .. [1] Javier Cerrillo and Jianshu Cao, Phys. Rev. Lett 112, 110401 (2014)
+    """
+
+    opt = {
+        "store_final_state": False,
+        "store_states": None,
+        "normalize_output": "ket",
+        "threshold": 0.0,
+        "num_learning": 0,
+    }
+    if options:
+        opt.update(options)
+
+    if not np.allclose(np.diff(times), times[1] - times[0]):
+        raise ValueError("The time should be uniformily distributed.")
+
+    if callable(dynmaps):
+        if not options["num_learning"]:
+            raise ValueError(
+                "When dynmaps is a callable, options['num_learning'] must be "
+                "the number of dynamical maps to compute."
+            )
+        dynmaps = [dynmaps(t) for t in times[: opt["num_learning"]]]
+
+    if not dynmaps[0].issuper:
+        raise ValueError("`dynmaps` entries must be super operators.")
```

Comment:
Is it worthwhile just checking that all dynmaps are super operators since it isn't expensive? We should perhaps also check that `dynmaps` isn't an empty list if that isn't allowed.